File: src/main/java/net/thedragonteam/thedragonlib/network/PacketSyncObject.java
package net.thedragonteam.thedragonlib.network;

import net.minecraft.client.Minecraft;
import net.minecraftforge.fml.common.network.simpleimpl.IMessage;
import net.minecraftforge.fml.common.network.simpleimpl.MessageContext;
import net.thedragonteam.thedragonlib.util.LogHelper;

public abstract class PacketSyncObject<REQ extends IMessage, REPLY extends IMessage> implements Runnable {

    public final REQ message;
    public final MessageContext ctx;
    public REPLY reply;

    public PacketSyncObject(REQ message, MessageContext ctx) {
        this.message = message;
        this.ctx = ctx;
    }

    @Override
    public abstract void run();

    public void addPacketServer() {
        switch (ctx.side) {
            case CLIENT:
                LogHelper.error("[SyncPacket#addPacketServer] HEY!!! I caught you this time! WRONG SIDE!!!! - " + message.getClass());
                return;
        }
        ctx.getServerHandler().player.getServer().addScheduledTask(this);
    }

    public void addPacketClient() {
        switch (ctx.side) {
            case SERVER:
                LogHelper.error("[SyncPacket#addPacketClient] HEY!!! I caught you this time! WRONG SIDE!!!! - " + message.getClass());
                return;
        }
        Minecraft.getMinecraft().addScheduledTask(this);
    }
}

File: src/main/java/net/thedragonteam/thedragonlib/network/PacketSyncObject.kt
package net.thedragonteam.thedragonlib.network

import net.minecraft.client.Minecraft
import net.minecraftforge.fml.common.network.simpleimpl.IMessage
import net.minecraftforge.fml.common.network.simpleimpl.MessageContext
import net.minecraftforge.fml.relauncher.Side
import net.thedragonteam.thedragonlib.util.LogHelper

abstract class PacketSyncObject<out REQ : IMessage, REPLY : IMessage>(val message: REQ, val ctx: MessageContext) : Runnable {
    var reply: REPLY? = null

    abstract override fun run()

    fun addPacketServer() {
        if (ctx.side == Side.CLIENT) {
            LogHelper.error("[SyncPacket#addPacketServer] HEY!!! I caught you this time! WRONG SIDE!!!! - " + message.javaClass)
            return
        }
        ctx.serverHandler.player.server!!.addScheduledTask(this)
    }

    fun addPacketClient() {
        if (ctx.side == Side.SERVER) {
            LogHelper.error("[SyncPacket#addPacketClient] HEY!!! I caught you this time! WRONG SIDE!!!! - " + message.javaClass)
            return
        }
        Minecraft.getMinecraft().addScheduledTask(this)
    }
}